File: DP/python/2294.py
from collections import deque
n,k = map(int,input().split())
coins = []
for _ in range(n):
  num = int(input())
  if num<=k:
    coins.append(num)
queue = deque()
max_num = 10**18
dp = [max_num for _ in range(k+1)]

for coin in coins:
  if coin>k: continue
  dp[coin] = 1
  queue.append(coin)
while queue:
  num = queue.popleft()
  for coin in coins:
    if coin+num >k:
      continue
    cnt = dp[coin+num]
    if (cnt == max_num):
      dp[coin+num] = dp[num]+1
      queue.append(coin+num)
    if coin+num == k:
      break
  if coin+num == k:
    break
if dp[k] == max_num:
  dp[k] = -1
print(dp[k])

File: DP/javascript/2294.js
const input = require("fs").readFileSync("/dev/stdin").toString().split("\n");
const [n, k] = input[0].split(" ").map(Number);
let coins = [];
let dp = new Array(k + 1);
let num,
  idx = 0;
let queue = [];
for (let i = 1; i < input.length - 1; i++) {
  num = Number(input[i]);
  if (num > k) continue;
  coins.push(num);
  dp[num] = 1;
  queue.push(num);
}

if (queue.length > 0) {
  while (queue.length > idx) {
    num = queue[idx++];
    for (coin of coins) {
      if (coin + num > k) continue;
      if (dp[coin + num] !== undefined) continue;
      dp[coin + num] = dp[num] + 1;
      if (coin + num === k) break;
      queue.push(coin + num);
    }
    if (coin + num === k) break;
  }
}
if (dp[k] === undefined) dp[k] = -1;
console.log(dp[k]);
